shader/svgf_temporal.comp

```diff
@@ -4,231 +4,181 @@
-layout(binding = 11, set = 0, rgba16f) uniform image2DArray svgf_color_hist;
-layout(binding = 12, set = 0, rgba16f) uniform image2DArray svgf_hist_specular;
-
-layout(push_constant) uniform push_constant_buffer
+layout(binding = 2, set = 0, rgba16f) uniform readonly image2DArray in_albedo;
+layout(binding = 3, set = 0, rg16_snorm) uniform readonly image2DArray previous_normal;
+layout(binding = 4, set = 0, rg32f) uniform readonly image2DArray in_screen_motion;
+layout(binding = 5, set = 0, rgba32f) uniform readonly image2DArray previous_color;
+layout(binding = 6, set = 0, rgba32f) uniform image2DArray out_color;
+layout(binding = 8, set = 0, rgba32f) uniform readonly image2DArray in_linear_depth;
+layout(binding = 9, set = 0, rgba32f) uniform readonly image2DArray previous_linear_depth;
+layout(binding = 10, set = 0, rgba32f) uniform readonly image2DArray previous_moments;
+layout(binding = 11, set = 0, rgba32f) uniform image2DArray out_moments;
+layout(binding = 12, set = 0, rgba32f) uniform readonly image2DArray in_diffuse;
+layout(binding = 13, set = 0, rgba32f) uniform readonly image2DArray previous_specular;
+layout(binding = 14, set = 0, rgba32f) uniform image2DArray out_specular;
+
+layout(binding = 7) buffer jitter_info_buffer
 {
-    ivec2 size;
-    int parity;
-    int level;
-    int stride;
-    int iteration_count;
-} control;
-
-vec3 get_specular(ivec3 p)
-{
-    return max(vec3(0.0), imageLoad(in_color, p).xyz - (imageLoad(in_albedo, p).xyz * imageLoad(in_diffuse, p).xyz));
-}
+    vec4 jitter[];
+} jitter_info;
 
-bool keep(ivec3 samplepos, vec3 current_normal, vec3 current_pos)
+bool temporal_reprojection_is_valid(ivec3 p, float z, float prev_z, vec3 normal, vec3 prev_normal, float motion_z)
 {
-    vec3 pos_diff = current_pos - imageLoad(previous_pos, samplepos).xyz;
-    vec3 normal = unpack_gbuffer_normal(imageLoad(previous_normal, samplepos).xy);
-
-#if 1
-    float diffsquared = dot(pos_diff, pos_diff);
-    float w;
-    if (diffsquared < 0.001)
-        w = 1.0;
-    else
-    {
-        vec3 tangent = normalize(pos_diff);
-        w = clamp(1.0f-abs(dot(tangent, current_normal))*1.0f, 0.0, 1.0) *
-            clamp(dot(normal, current_normal), 0.0f, 1.0f);
-    }
+    if (any(greaterThanEqual(p.xy, control.size)) || any(lessThan(p.xy, ivec2(0))))
+        return false;
+
+    if (isnan(z) || isnan(prev_z))
+        return false;
 
-    bool keep =
-    all(lessThan(samplepos.xy, control.size))
-    && all(greaterThanEqual(samplepos.xy, ivec2(0)))
-    && dot(normal, current_normal) > COS_LIMIT
-    && w > DISCARD_LIMIT;
-#else
-    bool keep =
-    all(lessThan(samplepos.xy, control.size))
-    && all(greaterThanEqual(samplepos.xy, ivec2(0)))
-    && dot(normal, current_normal) > COS_LIMIT
-    && dot(pos_diff, pos_diff) < SQRD_DIST_LIMIT;
-#endif
-
-    return keep;
+    if (abs(prev_z - motion_z) / z > 0.05)
+        return false;
+
+    if (dot(normal, prev_normal) < 0.5)
```

Comment:
Hmm. Maybe this function could return a float instead? Then, you could have a smooth "cutoff" for the normal condition. It should behave better with rotating objects.